pkg/catalog/cleanup.go

```diff
@@ -13,40 +13,34 @@ import (
 // where we allow plugins to be installed when target value is different even if target
 // values of “(empty), `global` and `kubernetes` can correspond to same root level command
 func DeleteIncorrectPluginEntriesFromCatalog() {
-	for _, c := range getCatalogs() {
-		plugins := c.List()
-		for i := range plugins {
+	c, unlock, err := getCatalogCache(true)
+	if err != nil {
+		return
+	}
+	defer unlock()
+
+	pluginAssociations := []PluginAssociation{c.StandAlonePlugins}
+	for _, spa := range c.ServerPlugins {
+		pluginAssociations = append(pluginAssociations, spa)
+	}
+
+	for i := range pluginAssociations {
+		for _, path := range pluginAssociations[i] {
+			pluginInfo, exists := c.IndexByPath[path]
+			if !exists {
+				continue
+			}
+
 			// The "unknown" target was previously used in two scenarios:
 			// 1- to represent the global target (>= v0.28 and < v0.90)
 			// 2- to represent either the global or kubernetes target (< v0.28)
 			// If we have a plugin with the "global" or "k8s" target we should remove any similar plugin using
 			// the "unknown" target.
-			if plugins[i].Target == configtypes.TargetGlobal || plugins[i].Target == configtypes.TargetK8s {
-				c.deleteOldTargetEntries(PluginNameTarget(plugins[i].Name, configtypes.TargetUnknown))
+			if pluginInfo.Target == configtypes.TargetGlobal || pluginInfo.Target == configtypes.TargetK8s {
+				delete(c.StandAlonePlugins, PluginNameTarget(pluginInfo.Name, configtypes.TargetUnknown))
```

Comment:
> Why is this changed to only delete the StandAlonePlugins section?
I'm not 100% but couldn't there be context-scoped plugins that have an entry using the old empty string to represent a k8s target?

Good catch. I should be using `pluginAssociations[i]` instead of `c.StandAlonePlugins`.


> Also, don't we also want to cleanup the IndexByPath and IndexByName sections as we used to do?
I think deleting IndexByPath and IndexByName can be considered a bug to me. I tried to explain the rationale behind that [here](https://github.com/vmware-tanzu/tanzu-cli/pull/472/files#r1308871509)